frontend/src/GameReportForm.tsx

```diff
@@ -65,93 +65,263 @@ const cities: Stronghold[] = [
     "Iron Hills (Fate of Erebor expansion only)",
 ];
 
-interface FormData {
-    winner: string | null;
-    loser: string | null;
-    side: Side | null;
-    victoryType: Victory | null;
-    matchType: Match | null;
-    competitionTypes: Competitive[];
-    league: League | null;
-    usedExpansions: boolean | null;
-    expansions: Expansion[];
-    wasTreebeardMustered: boolean | null;
-    usedHandicap: boolean | null;
-    actionTokens: number;
-    dwarvenRings: number;
-    gameTurns: number;
-    corruption: number;
-    didFellowshipReachMordor: boolean | null;
-    mordorTrack: number;
-    initialEyes: number;
-    wasAragornCrowned: boolean | null;
-    aragornCrownedTurn: number;
-    capturedStrongholds: Stronghold[];
-    interestRating: number;
-    comment: string;
+enum ErrorMessage {
+    Required = "Required",
+    OnSubmit = "Could not submit, please resolve errors",
+}
+
+export type FieldError = string | null;
+
+interface FieldData<T> {
+    value: T;
+    error: FieldError;
+    validate: () => FieldError;
+}
+
+export interface FormData {
+    winner: FieldData<string | null>;
+    loser: FieldData<string | null>;
+    side: FieldData<Side | null>;
+    victoryType: FieldData<Victory | null>;
+    matchType: FieldData<Match | null>;
+    competitionTypes: FieldData<Competitive[]>;
+    league: FieldData<League | null>;
+    usedExpansions: FieldData<boolean | null>;
+    expansions: FieldData<Expansion[]>;
+    wasTreebeardMustered: FieldData<boolean | null>;
+    usedHandicap: FieldData<boolean | null>;
+    actionTokens: FieldData<number>;
+    dwarvenRings: FieldData<number>;
+    gameTurns: FieldData<number>;
+    corruption: FieldData<number>;
+    didFellowshipReachMordor: FieldData<boolean | null>;
+    mordorTrack: FieldData<number>;
+    initialEyes: FieldData<number>;
+    wasAragornCrowned: FieldData<boolean | null>;
+    aragornCrownedTurn: FieldData<number>;
+    capturedStrongholds: FieldData<Stronghold[]>;
+    interestRating: FieldData<number>;
+    comment: FieldData<string | null>;
 }
 
 const UNINITIALIZED = -1;
 const INFINITE = 100;
 
 const initialFormData: FormData = {
-    winner: null,
-    loser: null,
-    side: null,
-    victoryType: null,
-    matchType: null,
-    competitionTypes: [],
-    league: null,
-    usedExpansions: null,
-    expansions: [],
-    wasTreebeardMustered: null,
-    usedHandicap: null,
-    actionTokens: 0,
-    dwarvenRings: 0,
-    gameTurns: UNINITIALIZED,
-    corruption: UNINITIALIZED,
-    didFellowshipReachMordor: null,
-    mordorTrack: UNINITIALIZED,
-    initialEyes: UNINITIALIZED,
-    wasAragornCrowned: null,
-    aragornCrownedTurn: UNINITIALIZED,
-    capturedStrongholds: [],
-    interestRating: -1,
-    comment: "",
+    winner: {
```

Comment:
The change to `FormData` even on its own adds a silly number of lines - in the next PR I want to split things off `GameReportForm` into smaller files, but I figured if I did it now it would be harder to review the diff.

Reply:
Like the idea of conceptual overhead 💯